### PR TITLE
Replace dependency on broken node-unzipper with native zlib

### PR DESCRIPTION
**Pull request by Durisvk**
### Description

Fixes:
    https://github.com/firebase/firebase-tools/issues/5614
    https://github.com/firebase/firebase-tools/issues/5677
    https://github.com/firebase/firebase-tools-ui/issues/939
    https://github.com/firebase/firebase-tools-ui/issues/940
    https://github.com/firebase/firebase-tools-ui/issues/942

The [node-unzipper](https://github.com/ZJONSSON/node-unzipper) package is generating incorrect files when unpacking the `ui` and `pubsub` emulator archive files (https://github.com/ZJONSSON/node-unzipper/issues/271). It generates a Syntax error and the emulator fails to execute. The issue happens on the latest NodeJS 18 version.

### Scenarios Tested

Automated: src/test/unzip.spec.ts

Manual:

Running the command before the fix

`firebase emulators:start --only database,auth,ui --import ./firebase-emulator --export-on-exit`

Output:
```
i  emulators: Starting emulators: auth, database
⚠  ui: Not starting the ui emulator, make sure you have run firebase init.
i  database: Database Emulator logging to database-debug.log
i  database: Importing data from /Users/jurajcarnogursky/ooo/vette/monorepo/firebase-emulator/database_export/vette-dev-default-rtdb.json
i  database: Importing data from /Users/jurajcarnogursky/ooo/vette/monorepo/firebase-emulator/database_export/vette-staging-default-rtdb.json
i  auth: Importing config from /Users/jurajcarnogursky/ooo/vette/monorepo/firebase-emulator/auth_export/config.json
i  auth: Importing accounts from /Users/jurajcarnogursky/ooo/vette/monorepo/firebase-emulator/auth_export/accounts.json
i  ui: Emulator UI logging to ui-debug.log
⚠  ui: Fatal error occurred: 
   Emulator UI has exited with code: 1, 
   stopping all running emulators
i  ui: Stopping Emulator UI
⚠  ui: Error stopping Emulator UI
i  database: Stopping Database Emulator
i  auth: Stopping Authentication Emulator
i  hub: Stopping emulator hub
i  logging: Stopping Logging Emulator
```

Then running `rm -rf ~/.cache/firebase/emulators`, linking my new version of `firebase-tools` and rerunning the same command again:

`firebase emulators:start --only database,auth,ui --import ./firebase-emulator --export-on-exit`

Output:
```
i  emulators: Starting emulators: auth, database
⚠  ui: Not starting the ui emulator, make sure you have run firebase init.
i  database: downloading firebase-database-emulator-v4.11.0.jar...
Progress: ==================================================================================================================================================================================================================================> (100% of 35MB
i  database: Database Emulator logging to database-debug.log
i  database: Importing data from /Users/jurajcarnogursky/ooo/vette/monorepo/firebase-emulator/database_export/vette-dev-default-rtdb.json
i  database: Importing data from /Users/jurajcarnogursky/ooo/vette/monorepo/firebase-emulator/database_export/vette-staging-default-rtdb.json
i  auth: Importing config from /Users/jurajcarnogursky/ooo/vette/monorepo/firebase-emulator/auth_export/config.json
i  auth: Importing accounts from /Users/jurajcarnogursky/ooo/vette/monorepo/firebase-emulator/auth_export/accounts.json
i  ui: downloading ui-v1.11.5.zip...
Progress: ===================================================================================================================================================================================================================================> (100% of 4MB
i  ui: Emulator UI logging to ui-debug.log

┌─────────────────────────────────────────────────────────────┐
│ ✔  All emulators ready! It is now safe to connect your app. │
│ i  View Emulator UI at http://127.0.0.1:5002/               │
└─────────────────────────────────────────────────────────────┘

┌────────────────┬────────────────┬────────────────────────────────┐
│ Emulator       │ Host:Port      │ View in Emulator UI            │
├────────────────┼────────────────┼────────────────────────────────┤
│ Authentication │ 127.0.0.1:9099 │ http://127.0.0.1:5002/auth     │
├────────────────┼────────────────┼────────────────────────────────┤
│ Database       │ 127.0.0.1:9000 │ http://127.0.0.1:5002/database │
└────────────────┴────────────────┴────────────────────────────────┘
  Emulator Hub running at 127.0.0.1:4400
  Other reserved ports: 4500

Issues? Report them at https://github.com/firebase/firebase-tools/issues and attach the *-debug.log files.
```

### Sample Commands

N/A

# IMPORTANT ❗ 

Is it possible to automatically execute `rm -rf ~/.cache/firebase` for users after updating the package? If not then the fix will just not work because the corrupted cached version will be preferred.

One workaround I can think of is introducing a cache versioning as a text file in the `~/.cache/firebase` folder. Whenever that number doesn't match the sourcecode or the file doesn't exist the cache is forcefully regenerated and the cache version is adjusted to match the one in the sourcecode.

### CODE REVIEW

File: src/test/fixtures/zip-files/node-unzipper-testData/compressed-standard/broken.zip
Comment (christhompsongoogle) [inline src/test/fixtures/zip-files/node-unzipper-testData/compressed-standard/broken.zip:1]: What is this broken.zip file for?

**Comment by christhompsongoogle** [comment]
> Is it possible to automatically execute rm -rf ~/.cache/firebase for users after updating the package? If not then the fix will just not work because the corrupted cached version will be preferred.

We can force a re-download of the corrupted emulators in the cache by bumping their versions.

File: src/test/fixtures/zip-files/node-unzipper-testData/compressed-standard/broken.zip
Comment (Durisvk) [reply src/test/fixtures/zip-files/node-unzipper-testData/compressed-standard/broken.zip:1]: sorry, it was copied over from the fixtures in `unzipper` and forgot to delete it since it's not used in the tests. Going to remove it.

I've also included LICENSE file from node-unzipper package since I added some of the fixtures from there - https://github.com/ZJONSSON/node-unzipper/blob/master/LICENSE . I am not an expert on license policies - is it okay to keep it in there or should I remove it?

**Comment by christhompsongoogle** [comment]
> awesome, @christhompsongoogle , in case this gets released can you keep track of bumping the version of ui emulator and pubsub emulator?

Sure, I'd be happy to bump the versions once this is in

**Comment by christhompsongoogle** [comment]
After doing some testing on Windows it seems that there's an issue with the mkdir:

```
i  ui: downloading ui-v1.11.5.zip... {"metadata":{"emulator":{"name":"ui"},"message":"downloading ui-v1.11.5.zip..."}}
[2023-04-25T02:11:36.192Z] >>> [apiv2][query] GET https://storage.googleapis.com/firebase-preview-drop/emulator/ui-v1.11.5.zip
[2023-04-25T02:11:36.220Z] <<< [apiv2][status] GET https://storage.googleapis.com/firebase-preview-drop/emulator/ui-v1.11.5.zip 200
[2023-04-25T02:11:36.221Z] <<< [apiv2][body] GET https://storage.googleapis.com/firebase-preview-drop/emulator/ui-v1.11.5.zip [stream]
Progress: ==============================================================================================> (100% of 4MB)
i  emulators: Shutting down emulators. {"metadata":{"emulator":{"name":"hub"},"message":"Shutting down emulators."}}
i  ui: Stopping Emulator UI {"metadata":{"emulator":{"name":"ui"},"message":"Stopping Emulator UI"}}
i  database: Stopping Database Emulator {"metadata":{"emulator":{"name":"database"},"message":"Stopping Database Emulator"}}
!  Database Emulator has exited upon receiving signal: SIGINT
i  firestore: Stopping Firestore Emulator {"metadata":{"emulator":{"name":"firestore"},"message":"Stopping Firestore Emulator"}}
!  Firestore Emulator has exited upon receiving signal: SIGINT
i  auth: Stopping Authentication Emulator {"metadata":{"emulator":{"name":"auth"},"message":"Stopping Authentication Emulator"}}
i  storage: Stopping Storage Emulator {"metadata":{"emulator":{"name":"storage"},"message":"Stopping Storage Emulator"}}
i  hub: Stopping emulator hub {"metadata":{"emulator":{"name":"hub"},"message":"Stopping emulator hub"}}
i  logging: Stopping Logging Emulator {"metadata":{"emulator":{"name":"logging"},"message":"Stopping Logging Emulator"}}
[2023-04-25T02:11:36.443Z] Error: ENOENT: no such file or directory, mkdir ''
```

File: src/test/unzip.spec.ts
Comment (joehan) [inline src/test/unzip.spec.ts:1]: This thorough testing is great! However, these seem like integration tests to me. They are longer running than is ideal for our unit tests, and they also occasionally flake due to timeouts when run locally. 

I think these tests would be appropriate as integration tests. Could you separate them into a different npm command `test:unzip`, and ensure that they are not run as part of `npm run test` ?

File: src/test/unzip.spec.ts
Comment (joehan) [inline src/test/unzip.spec.ts:46]: This seems like it should not be commented out

File: src/test/unzip.spec.ts
Comment (joehan) [inline src/test/unzip.spec.ts:19]: Unclear to me why we need this needs to be set up as a self executing function. Could this just be:
```
before(async () => {
 ...
}

```

File: src/test/unzip.spec.ts
Comment (joehan) [inline src/test/unzip.spec.ts:50]: Can remove this linter violation by refactoring to:
```
it("should unzip...", async () => {}).timeout(2000);
```

File: src/test/unzip.spec.ts
Comment (joehan) [inline src/test/unzip.spec.ts:66]: Can remove this linter violation by refactoring to:

it("should unzip...", async () => {}).timeout(2000);

**Review by joehan — CHANGES_REQUESTED** [review]
Really appreciate the contribution! Some changes to make, but this is a great start

**Comment by Durisvk** [comment]
> outputFilePath

Hi, unfortunately I have no Windows device, I've pushed a fix replacing `/` with `path.sep` which should match the operating system, could you please retry?

**Comment by Durisvk** [comment]
Hi @joehan, thank you for the great points. I've addressed them.

Regarding moving the integration tests for unzip into a separate test - is it okay that I've moved them under `npm test:emulator` command since the logic is used within the emulators?

**Comment by christhompsongoogle** [comment]
> could you please retry?

Here's the latest issue I encountered
```
Progress: ==============================================================================================> (100% of 4MB)
i  emulators: Shutting down emulators. {"metadata":{"emulator":{"name":"hub"},"message":"Shutting down emulators."}}
i  ui: Stopping Emulator UI {"metadata":{"emulator":{"name":"ui"},"message":"Stopping Emulator UI"}}
i  database: Stopping Database Emulator {"metadata":{"emulator":{"name":"database"},"message":"Stopping Database Emulator"}}
!  Database Emulator has exited upon receiving signal: SIGINT
i  firestore: Stopping Firestore Emulator {"metadata":{"emulator":{"name":"firestore"},"message":"Stopping Firestore Emulator"}}
!  Firestore Emulator has exited upon receiving signal: SIGINT
i  auth: Stopping Authentication Emulator {"metadata":{"emulator":{"name":"auth"},"message":"Stopping Authentication Emulator"}}
i  storage: Stopping Storage Emulator {"metadata":{"emulator":{"name":"storage"},"message":"Stopping Storage Emulator"}}
i  hub: Stopping emulator hub {"metadata":{"emulator":{"name":"hub"},"message":"Stopping emulator hub"}}
i  logging: Stopping Logging Emulator {"metadata":{"emulator":{"name":"logging"},"message":"Stopping Logging Emulator"}}
[2023-04-26T21:46:51.161Z] Error: EISDIR: illegal operation on a directory, open 'C:\Users\chris\.cache\firebase\emulators\ui-v1.11.5\client'
```

The directory has a client directory but it's empty - and no server directory exists there

**Comment by Durisvk** [comment]
Hey @christhompsongoogle, sorry for the trouble. I've pushed a new version with potential fix and additional console logs in case the fix doesn't work.

In case you have time, can you test it? I'll have my brother over at my place on Saturday so I could borrow his Windows laptop to debug this but currently I don't have access to any windows machine.

File: scripts/emulator-tests/functions/index.js
Comment (joehan) [inline scripts/emulator-tests/functions/index.js:1]: Could you run `npm run format`? This file is failing whitespace linting

File: src/unzip.ts
Comment (joehan) [inline src/unzip.ts:63]: Please switch this to logger.debug(). That way, the logs will only appear in --debug mode.

File: src/unzip.ts
Comment (joehan) [inline src/unzip.ts:53]: Please switch this to logger.debug(). That way, the logs will only appear in --debug mode.

**Review by joehan — COMMENTED** [review]
A few more minor tweaks, but this is very close to LGTM.

**Comment by Durisvk** [comment]
@joehan I've adjusted the code based on your comments. I've also rebased on the current master.

@christhompsongoogle, once you'll be retesting this, can you include a `--debug` flag? (I've changed console.logs to logger.debugs)

**Review by joehan — APPROVED** [review]
LGTM once @christhompsongoogle confirms that it looks good on Windows

**Comment by christhompsongoogle** [comment]
> can you test it

Just rested on Windows and everything looks great. Thank you for the windows fixes !

LGTM on this one

**Review by christhompsongoogle — APPROVED** [review]

**Comment by Durisvk** [comment]
Hey should I remove the unnecessary debug logs? in case someone wants to debug something so that they don't see these unnecessary logs.

**Comment by christhompsongoogle** [comment]
> Hey should I remove the unnecessary debug logs? in case someone wants to debug something so that they don't see these unnecessary logs.

I don't mind leaving them in - they'll require some formatting though.

**Comment by Durisvk** [comment]
> > Hey should I remove the unnecessary debug logs? in case someone wants to debug something so that they don't see these unnecessary logs.
> 
> I don't mind leaving them in - they'll require some formatting to satisfy lint before we merge.

I'm having a really tough time understanding the linting rules (there's plenty of warnings popping up). Can you advise what needs to be changed before merging?

**Comment by joehan** [comment]
Hey @Durisvk, I did some more testing this weekend and realized that the extensions emulator isn't working under this PR. I'm going to try to debug/fix it when I have some time this week, but if you want to take a crack at it, the easiest way to exercise the path is the integration test:

FBTOOLS_TARGET_PROJECT=<your project id> npm run test:extensions-emulator

**Comment by joehan** [comment]
So, it seems like this implementation doesn't handle zips that have the data descriptor bit set: https://en.wikipedia.org/wiki/ZIP_(file_format)#:~:text=the%20compressed%20data.-,Data%20descriptor,-%5Bedit%5D.

Incidentally, Extensions zips all have this bit set. Looking into how to add support for this now

File: src/unzip.ts
Comment (christhompsongoogle) [inline src/unzip.ts:99]: I don't love that we're using the compressedSize to make decisions - though that might legitimately be part of the header. Do you have a link to the header breakdown (or the bitFlag === 8 details) we could leave as a comment?

**Review by christhompsongoogle — APPROVED** [review]

File: src/unzip.ts
Comment (joehan) [reply src/unzip.ts:99]: Good catch - this actually does not need to be here, was leftover from a different approach i tried.

**Comment by christhompsongoogle** [comment]
> @Durisvk @christhompsongoogle @joehan do you have a timeline when this fix will be available in the official NPM? Thanks!

This should be available now, with firebase-tools v12.0 (and possibly 11.30 before that)

**Pull request merged**
